### PR TITLE
Handle job cancellation for root span

### PR DESCRIPTION
**Pull request by huaiyan**
# Description
1. Record KeyboardInterrupt exception for span (otlp's code doesn't handle BaseException but only Exception)
2. Raise KeyboardInterrupt in FlowNodesScheduler
3. Remove `with ThreadPoolExecutor` to guarantee KeyboardInterrupt could be received by outside.

Why make this change: **we want to support cancellation for span.**

We have 2 schedulers, one is FlowNodesScheduler, another is AsyncNodesScheduler.

For most cases, we are using FlowNodesScheduler.
In FlowNodesScheduler, there is `with ThreadPoolExecutor`, which will wait all running threads finish, no matter what exception is raised. That's why our cancellation is blocked.
image

So, in this PR, remove the `with ThreadPoolExecutor` to make the KeyboardInterrupt could be received by outside.

Why also change other tests:
For now, our exception could be received by outside immediately, the exception type and execution time are affected.

For test_batch_timeout, situation is very complex.
The test set 5 seconds batch run timeout and 600 seconds line run timeout.
But actually, for batch run execution, line run timeout is calculated by batch run timeout. So, 600 seconds won't take effect, the real line run timeout is less than 5 seconds.
But before this change, even we raise line run timeout exception, we can't exit ThreadPoolExecutor to finish line run.
From process perspective, the line run has never finish, so assign one `BatchExecutionTimeoutError`.
After this change, line run timeout exception could be raised out soon, so the behavior is changed.

# All Promptflow Contribution checklist:
- [ ] **The pull request does not introduce [breaking changes].**
- [ ] **CHANGELOG is updated for new features, bug fixes or other significant changes.**
- [ ] **I have read the [contribution guidelines](../CONTRIBUTING.md).**
- [ ] **Create an issue and link to the pull request to get dedicated review from promptflow team. Learn more: [suggested workflow](../CONTRIBUTING.md#suggested-workflow).**

## General Guidelines and Best Practices
- [ ] Title of the pull request is clear and informative.
- [ ] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/master/documentation/development-docs/cleaning-up-commits.md).

### Testing Guidelines
- [ ] Pull request includes test coverage for the included changes.

### CODE REVIEW

File: src/promptflow-core/promptflow/executor/flow_executor.py [hunk 1]
```diff
@@ -847,7 +847,7 @@ def _exec_inner_with_trace(
     ):
         # need to get everytime to ensure tracer is latest
         otel_tracer = otel_trace.get_tracer("promptflow")
-        with otel_tracer.start_as_current_span(self._flow.name) as span:
+        with otel_tracer.start_as_current_span(self._flow.name) as span, self._record_keyboard_interrupt_to_span(span):
```
Comment (zzn2) [inline src/promptflow-core/promptflow/executor/flow_executor.py:850]: Do we do this only for "Flow" spans? How about a flex flow that does not have a "Flow" span?

Comment (huaiyan) [reply src/promptflow-core/promptflow/executor/flow_executor.py:850]: Good catch. Maybe open another PR for flex scenario.
This PR is enough dangerous.

Comment (huaiyan) [reply src/promptflow-core/promptflow/executor/flow_executor.py:850]: Another PR: https://github.com/microsoft/promptflow/pull/2918

File: src/promptflow-core/promptflow/executor/flow_executor.py [hunk 2]
```diff
@@ -874,6 +874,16 @@ def _exec_inner_with_trace(
...
             return output, aggregation_inputs
 
+    @contextlib.contextmanager
+    def _record_keyboard_interrupt_to_span(self, span: Span):
+        try:
+            yield
+        except KeyboardInterrupt as ex:
+            if span.is_recording():
+                span.record_exception(ex)
+                span.set_status(StatusCode.ERROR, "KeyboardInterrupt received, execution cancelled.")
```
Comment (guming-learning) [inline src/promptflow-core/promptflow/executor/flow_executor.py:884]: KeyboardInterrupt is our internal implementation for all cases except when user actually hit ctrl-c locally.
The message should just be "Excecution cancelled."

File: src/promptflow-core/promptflow/executor/_flow_nodes_scheduler.py [hunk 3]
```diff
@@ -51,45 +81,56 @@ def execute(
...
+        if threading.current_thread() is threading.main_thread():
+            signal.signal(signal.SIGINT, signal_handler)
+            signal.signal(signal.SIGTERM, signal_handler)
+        else:
+            flow_logger.info(
+                "Current thread is not main thread, skip signal handler registration in AsyncNodesScheduler."
+            )
         parent_context = contextvars.copy_context()
-        with ThreadPoolExecutor(
+        # If we use with ThreadPoolExecutor, we can't use KeyboardInterrupt to exit waiting.
```
Comment (thy09) [inline src/promptflow-core/promptflow/executor/_flow_nodes_scheduler.py:92]: Add more comments to explain the logic here.

File: src/promptflow-core/promptflow/executor/_flow_nodes_scheduler.py [hunk 4]
```diff
@@ -23,6 +27,32 @@
...
+    # Specifically, it does not exit non-daemon thread
+    # os._exit https://docs.python.org/3/library/os.html#os._exit
+    # Exit the process with status n, without calling cleanup handlers, flushing stdio buffers, etc.
+    # Specifically, it stops process without waiting for non-daemon thread.
+    os._exit(0)
+
+
+def signal_handler(sig, frame):
+    """
+    We have no way to cancel running tasks in ThreadPoolExecutor, so we raise KeyboardInterrupt to exit the process.
```
Comment (thy09) [inline src/promptflow-core/promptflow/executor/_flow_nodes_scheduler.py:47]: Add more detail about why
1 we have a separate thread here
2 raise an exception here

And explain what if we do not do this?

File: src/promptflow-core/promptflow/executor/_flow_nodes_scheduler.py [hunk 5]
```diff
@@ -51,45 +81,56 @@ def execute(
...
+                for each_future in completed_futures:
+                    del self._future_to_node[each_future]
+                if timeout_task and timeout_task.done():
+                    raise LineExecutionTimeoutError(self._context._line_number, line_timeout_sec)
+                self._execute_nodes(executor)
+            event.set()
+            executor.shutdown()
+        except Exception as e:
+            # When meet exception, mark event to exit timeout task.
+            event.set()
```
Comment (guming-learning) [inline src/promptflow-core/promptflow/executor/_flow_nodes_scheduler.py:122]: Why not shutdown here?

Comment (huaiyan) [reply src/promptflow-core/promptflow/executor/_flow_nodes_scheduler.py:122]: If call shudown here, will wait for running threads finish.
Maybe add shudown(wait=False) is anyway better than current code?

File: src/promptflow-core/promptflow/executor/_flow_nodes_scheduler.py (hunk 4 above)
Comment (huaiyan) [reply src/promptflow-core/promptflow/executor/_flow_nodes_scheduler.py]: Fixed~

**Review by thy09 — APPROVED** [review]

**Review by guming-learning — APPROVED** [review]

**Pull request merged**